examples/cosmos/ics20/bank.qnt

```diff
@@ -0,0 +1,240 @@
+   * TYPES
+   * **************************************************************************/
+
+  type Address = str
+  type Denomination = str
+  type Amount = int
+
+  type TokenBalances = Denomination -> Amount
+  type Accounts = Address -> TokenBalances
+
+  /* ***************************************************************************
+   * FUNCTIONAL LAYER, manipulating balances
+   * **************************************************************************/
+
+  // Get balances of all denominations in the given account `addr`.
+  pure def getBalances(accounts: Accounts, addr: Address) : TokenBalances = {
+    if (accounts.keys().contains(addr)) {
+      accounts.get(addr)
+    } else {
+      Map()
+    }
+  }
+
+  // Get the amount of tokens of denomination `denom` in `balances`.
+  pure def getBalance(balances: TokenBalances, denom: Denomination): Amount = {
+    if (balances.keys().contains(denom)) balances.get(denom) else 0
+  }
+
+  /* Return `accounts`, where `address` is updated to hold
+   * `currentAmount + amount` tokens of denomination `denom`.
+   * 
+   * `currentAmount` refers to the number of tokens of denomination `denom` held
+   * by `address` in `accounts`.
+   *
+   * If there currently is no account under `address`, return accounts including
+   * `address` with exactly `amount`-many `denom` tokens.
+   */
+  pure def manipulateAccount(accounts: Accounts, address: Address,
+                             denom: Denomination, amount: Amount) : Accounts = {
+    val accountCoins = accounts.getBalances(address)
+    val newCoins = accountCoins.put(denom, accountCoins.getBalance(denom) + amount)
+    accounts.put(address, newCoins)
+  }
+
+  /* Return `accounts`, where `amount`-many tokens of denomination `denom`
+   * vanished out of `sender`'s account into thin air.
+   *
+   * This intentionally does NOT check whether `amount > 0`.
+   */
+  pure def BurnCoins(accounts: Accounts, sender: Address,
+                     denom: Denomination, amount: Amount): Accounts = {
+    manipulateAccount(accounts, sender, denom, -amount)
+  }
+
+  /* Return `accounts`, where `amount`-many tokens of denomination `denom` were
+   * added to `sender`'s account out of thin air.
+   *
+   * This intentionally does NOT check whether `amount > 0`.
+   */
+  pure def MintCoins(accounts: Accounts, receiver: Address,
+                     denom: Denomination, amount: Amount): Accounts = {
+    manipulateAccount(accounts, receiver, denom, amount)
+  }
+
+  /* Return `state`, where `amount`-many tokens of denomination `denom` were
+   * moved from `sender`'s account into `receiver`'s acount.
+   *
+   * This intentionally does NOT check whether `amount > 0`.
+   */
+  pure def TransferCoins(accounts: Accounts, sender: Address, receiver: Address,
+                         denom: Denomination, amount: Amount): Accounts = {
+    accounts
+      .BurnCoins(sender, denom, amount)
+      .MintCoins(receiver, denom, amount)
```

Comment:
That's a very interesting way of expressing a transfer. I have never seen anything like that before, but it makes a lot of sense!

Reply:
I'm not a big fan of this. Even though it works and makes sense at this specification's abstraction level, it might be misleading to people using this spec to understand how transfers work. Since this isn't specially clearer or significantly more concise than writting:
```suggestion
      .addToBalance(sender, denom, -amount)
      .addToBalance(receiver, denom, amount)
```

I'd avoid using `mint` and `burn` abstractions for expressing it.

This might me too much of a personal preference here, and even if this actually is confusing to people, this shouldn't be a reference for bank module implementations anyhow. So I'm fine either way :smile:

Reply:
I was undecided about this, so let's go with @bugarela's suggestion.